### PR TITLE
[boschindego] Implement OAuth2 authorization

### LINKED ISSUE #14536: [boschindego] Broken since cloud upgrade
On February 22nd 2023 Indego users received an e-mail from Bosch announcing a cloud upgrade and days of downtime. This e-mail mentioned also a migration to Bosch ID account and a requirement to upgrade to Bosch Smart Gardening app version 4.0. The need to upgrade the app clearly indicated to me some kind of upcoming breaking change:
https://community.openhab.org/t/boschindego-bosch-indego-binding-not-working-any-more/135825/8

Today I migrated my account to Bosch ID and the new version of the app was able to connect again.

The binding however receives **ECONNREFUSED**.

In my initial investigation I downloaded the new APK for Android and decompiled it using dex2jar. With JD GUI I found the new endpoint: https://api.indego-cloud.iot.bosch-si.com/

Unfortunately not only the endpoint changed. So far I have received nothing other than HTTP code 401/unauthorized. I found also references to known paths like "api/v1/alms/" so I don't think the entire API changed, but I do think they changed the way to authenticate when they migrated to Bosch ID.

Full e-mail follows:

## Cloud Upgrade: Downtime Bosch Smart Gardening App
Due  to an upgrade in our central cloud system, the Bosch Smart Gardening  App and connected services that are used for Bosch Indego will be  unavailable from Monday 27th of February to Wednesday 1st of March 2023.

Following  this upgrade, all users are required to update the Bosch Smart  Gardening App to version 4.0. Previous versions will no longer provide  access to the Indego.

As  part of the upgrade of our central cloud system, we are also going to  move from using local email accounts to central Bosch ID accounts.This  will allow our users to use their account for all Bosch products,  including our MyBosch platform for further service advantages around  your product like warranty extensions.

Depending  on the type of user account the Indego is paired to, there might be  some additional steps that need to be taken for this upgrade. Details  can be seen below.

### Email accounts

To  continue to access the Bosch Smart Gardening App, users need to create a  Bosch ID account with the same email address that has been used on the  previous email account.

The steps to create a new Bosch ID account are as follows:
- Select the button ‘Bosch ID’ when logging in to the updated App.
- This will redirect the user to the Bosch ID login window.
- Create a Bosch ID account by selecting ‘Not registered yet’.
- Confirm the account creation via the link sent by email.
- Log in to the Bosch Smart Gardening App using the login details of the newly created Bosch ID account.

If there is already a Bosch ID account existing with the same email address, this account can directly be used to access the paired Indego by simply logging in to the existing Bosch ID account.

It is important that users create their Bosch ID account using the same email address that has been previously used for the creation of their Indego account. 

This will ensure that the paired Indego will automatically be transferred to the new Bosch ID account.

### Bosch ID, Facebook, Apple ID logins

If one of these options has been used in the past, please select the respective button, and follow the instructions in the app.

Please note that home automation services (e.g. Amazon Alexa, Google Home, IFTTT) cannot be used to operate the Indego until March 2023 when a further upgrade will be completed for these tools to be re-enabled.

After all changes have been completed, skills that have been used before must be linked again with user accounts to continue using them.

As part of our commitment to providing and maintaining high quality products and services, we have taken all necessary measures to ensure a smooth upgrade before the start of the 2023 mowing season.

We would further like to inform our users that Bosch ID will soon become SingleKey ID. Users do not need to take further action for the transfer to SingleKey ID. Once the transfer is concluded users will be able to use their existing Bosch ID credentials to login to SingleKey ID.

For further information please contact Indego.Support@de.bosch.com.

### Q&A

**Are all connected mowers affected by this temporary downtime of the Bosch Gardening App?** 

Yes. The upgrade of our system will affect all connected mowers. Therefore, all Bosch Smart Gardening App users will be affected by this three-day downtime. 

**Why will connected services like Amazon Alexa be unavailable for a longer time?**

Due to the way our system is connected with home automation services, such as Amazon Alexa, we have to complete the upgrade in multiple stages. We have focused on completing the work on the Bosch Smart Gardening App first. Following the first upgrade, we will apply another upgrade that will take place as soon as possible to provide all functionalities related to home automation services.

**Why does my account need to be transferred to Bosch ID?**

We are aiming to improve the user experience by providing a single point of access to a wide range of our digital services.

To achieve this, we are discontinuing the use of local login solutions and moving to the central Bosch ID account solution. 

**What if I would rather login with Facebook or Apple ID instead of using Bosch ID?**

We will only be able to automatically transfer the paired Indego from email accounts to Bosch ID accounts. An automatic transfer to a Facebook or Apple ID login is not possible.

If users wish to use their Facebook or Apple ID login details in future, the Indego needs to be unpaired from the existing account first. This will require the initial creation of a Bosch ID account to access and unpair the Indego.

Upon experiencing issues or needing assistance, please contact the Indego Support Team.

When contacting the Indego Support Team, please provide the email address the Indego is paired to as well as the 9-digit serial number, which can be found on the rating label on the back of the Indego.

**What if I create the new Bosch ID account with a different email address?** 

We will only be able to automatically transfer the paired Indego from email accounts to Bosch ID accounts if the email address matches. 
An automatic transfer to an account using a different email address is not possible. 

If users wish to use a different email address in future, the Indego needs to be unpaired from the existing account first. This will require the initial creation of a Bosch ID account using the original email address to access and unpair the Indego. 

Upon experiencing issues or needing assistance, please contact the Indego Support Team.
  
When contacting the Indego Support Team, please provide the email address the Indego is paired to as well as the 9-digit serial number, which can be found on the rating label on the back of the Indego.

**Why will Bosch ID transfer to SingleKey ID?** 

While Bosch ID provided access to numerous websites, shops, apps, and services by Bosch, SingleKey ID will allow the access to digital services of a growing number of partners worldwide.

**Why can’t I continue to use the current version of the app after the upgrade?** 

The Bosch Smart Gardening App links to a number of other systems. As these systems will change, all users are required to update the Bosch Smart Gardening App to version 4.0 in order to connect with all relevant supporting systems. It will no longer be possible to gain access to the Indego using older app versions.

### PR DESCRIPTION
**Pull request by jlaur**
On March 1st 2023 Bosch migrated to central Bosch ID accounts. This has two direct consequences for the binding:
- The previous authentication mechanism has been replaced by OAuth2.
- Multiple mowers can now be assigned to the same Bosch ID account.

This pull request:
- Partially implements the OAuth2 authorization flow.
- Introduces a bridge to represent the API account holding OAuth2.
- Introduces discovery of devices.
- Requires each device to carry serial number configuration, where previously each device would have its own account.

Some manual steps are needed for upgrading:
- A bridge thing must be created.
- The device thing must have this bridge assigned and a serial number configured (already available as property).
- A console command must be executed to complete the OAuth authorization flow.

Keep in mind though that the binding has been fully broken since March 1st.

I'm still trying to find a (user-friendly) way to retrieve the OAuth2 authorization code. This seems difficult, at least for me. Therefore I'm publishing this PR as a first step to get the binding to work again.

The problem yet unresolved is the use of a client id extracted from the Android app with a hardcoded redirect URL "com.bosch.indegoconnect://login/?code=XXXXXXXX".

Testing:
- [org.openhab.binding.boschindego-3.4.3-SNAPSHOT.jar](https://drive.google.com/file/d/1_DnODsPKEiEyWEMde79LeFdo87aSANPR/view?usp=sharing)
- [org.openhab.binding.boschindego-4.0.0-SNAPSHOT.jar](https://drive.google.com/file/d/1DIYxQKdqgkEMQtcUkW7atNLy1T9hOsMO/view?usp=sharing)

Fixes #14536

### CODE REVIEW

**Comment by jlaur** [comment]
@openhab/add-ons-maintainers, if this PR would be reviewed and completed during the Easter, I should be able to quickly backport it to 3.4.x so it could be included in 3.4.3.

**Comment by jlaur** [comment]
My mower has been authenticated for more than a week now without any issues. I don't have any plans for the near future to improve the cumbersome authorization flow described in the README.

**Comment by jlaur** [comment]
@openhab/add-ons-maintainers - since the binding is currently broken by cloud changes, please consider this PR for the upcoming milestone.

File: bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschAccountHandler.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    private OAuthClientService oAuthClientService;
+    private IndegoController controller;
+
+    public BoschAccountHandler(Bridge bridge, HttpClient httpClient, OAuthFactory oAuthFactory) {
+        super(bridge);
+
+        this.oAuthFactory = oAuthFactory;
+
+        oAuthClientService = oAuthFactory.createOAuthClientService(this.getThing().getUID().getAsString(),
```
Comment (wborn) [inline bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschAccountHandler.java:62]:
```suggestion
        oAuthClientService = oAuthFactory.createOAuthClientService(getThing().getUID().getAsString(),
```

File: bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschAccountHandler.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            } catch (IOException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                        "@text/offline.comm-error.oauth2-authorization-failed");
+            }
+        });
+    }
+
+    @Override
+    public void dispose() {
+        this.oAuthFactory.ungetOAuthService(this.getThing().getUID().getAsString());
```
Comment (wborn) [inline bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschAccountHandler.java:92]:
```suggestion
        oAuthFactory.ungetOAuthService(getThing().getUID().getAsString());
```

File: bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschIndegoHandler.java [hunk 3]
```diff
@@ -109,41 +113,56 @@ public BoschIndegoHandler(Thing thing, HttpClient httpClient, BoschIndegoTransla
...
         stateActiveRefreshIntervalSeconds = (int) config.stateActiveRefresh;
-        String username = config.username;
-        String password = config.password;
 
-        if (username == null || username.isBlank()) {
+        Bridge bridge = getBridge();
+        if (bridge == null) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
-                    "@text/offline.conf-error.missing-username");
+                    "@text/offline.conf-error.missing-brige");
```
Comment (wborn) [inline bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschIndegoHandler.java:123]:
```suggestion
                    "@text/offline.conf-error.missing-bridge");
```

File: bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschIndegoHandler.java [hunk 4]
```diff
@@ -109,41 +113,56 @@ public BoschIndegoHandler(Thing thing, HttpClient httpClient, BoschIndegoTransla
...
         }
-        if (password == null || password.isBlank()) {
+
+        ThingHandler handler = bridge.getHandler();
+        if (handler instanceof BoschAccountHandler accountHandler) {
+            this.oAuthClientService = accountHandler.getOAuthClientService();
+        } else {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
-                    "@text/offline.conf-error.missing-password");
+                    "@text/offline.conf-error.missing-brige");
```
Comment (wborn) [inline bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschIndegoHandler.java:132]:
```suggestion
                    "@text/offline.conf-error.missing-bridge");
```

File: bundles/org.openhab.binding.boschindego/src/main/resources/OH-INF/i18n/boschindego.properties [hunk 5]
```diff
@@ -53,10 +53,11 @@ channel-type.boschindego.textualstate.label = Textual State
 
 # thing status descriptions
 
-offline.comm-error.authentication-failure = The login credentials are wrong or another client is connected to your Indego account
+offline.conf-error.missing-brige = No bridge configured
```
Comment (wborn) [inline bundles/org.openhab.binding.boschindego/src/main/resources/OH-INF/i18n/boschindego.properties:56]:
```suggestion
offline.conf-error.missing-bridge = No bridge configured
```

**Review by wborn — CHANGES_REQUESTED** [review]
Thanks for making it working again! I only have a few minor comments.

**Comment by jlaur** [comment]
@wborn - thanks for reviewing this. I have addressed your comments.

**Review by wborn — APPROVED** [review]
Very good! :+1:

**Comment by wborn** [comment]
Now you can also create a PR to add an upgrade notice. :wink:

**Comment by BillGOH** [comment]
@jlaur Thank you for your great work! First tests look very good.

**Pull request merged**